lib/Library.php

```diff
@@ -140,31 +140,49 @@ public static function GetEbooksByTag(string $tag): array{
 	}
 
 	/**
-	 * @return array<string, Collection>
+	 * @return array<Collection>
 	 * @throws Exceptions\AppException
 	 */
 	public static function GetEbookCollections(): array{
-		/** @var array<string, Collection> */
-		return self::GetFromApcu('collections');
+		$collections = Db::Query('
+					SELECT distinct UrlName, Name
```

Comment:
Is `distinct` required? It should not possible to have more than one record with the same `UrlName`.

Reply:
The `Collections` table is denormalized:

```
> select * from Collections limit 20;
+--------------+---------+----------------------------------------------------+-----------------------------------------------+----------------+--------+
| CollectionId | EbookId | Name                                               | UrlName                                       | SequenceNumber | Type   |
+--------------+---------+----------------------------------------------------+-----------------------------------------------+----------------+--------+
|         2072 |       3 | Sherlock Holmes                                    | sherlock-holmes                               |              3 | series |
|         2073 |       3 | Haycraft-Queen Cornerstones                        | haycraft-queen-cornerstones                   |           NULL | set    |
|         2074 |       5 | Scarlet Pimpernel                                  | scarlet-pimpernel                             |              9 | series |
|         2075 |       9 | Poirot                                             | poirot                                        |              6 | series |
|         2076 |      11 | The Guardian’s Best 100 Novels in English (2015)   | the-guardians-best-100-novels-in-english-2015 |             28 | set    |
|         2077 |      11 | The BBC’s 100 Greatest British Novels (2015)       | the-bbcs-100-greatest-british-novels-2015     |             52 | set    |
|         2078 |      13 | Chronicles of Carlingford                          | chronicles-of-carlingford                     |              4 | series |
|         2079 |      14 | Anne of Green Gables                               | anne-of-green-gables                          |              2 | series |
|         2080 |      17 | Sherlock Holmes                                    | sherlock-holmes                               |              8 | series |
|         2081 |      17 | Haycraft-Queen Cornerstones                        | haycraft-queen-cornerstones                   |           NULL | set    |
|         2082 |      18 | Voyages Extraordinaires                            | voyages-extraordinaires                       |             42 | series |
|         2083 |      19 | Modern Library’s 100 Best Novels                   | modern-librarys-100-best-novels               |             88 | set    |
|         2084 |      19 | The Guardian’s Best 100 Novels in English (2015)   | the-guardians-best-100-novels-in-english-2015 |             35 | set    |
|         2085 |      21 | Modern Library’s 100 Best Novels                   | modern-librarys-100-best-novels               |             32 | set    |
|         2086 |      21 | The Guardian’s Best 100 Novels in English (2015)   | the-guardians-best-100-novels-in-english-2015 |             36 | set    |
|         2087 |      24 | Haycraft-Queen Cornerstones                        | haycraft-queen-cornerstones                   |           NULL | set    |
|         2088 |      30 | Modern Library’s 100 Best Nonfiction               | modern-librarys-100-best-nonfiction           |             31 | set    |
|         2089 |      35 | Modern Library’s 100 Best Novels                   | modern-librarys-100-best-novels               |             69 | set    |
|         2090 |      38 | Modern Library’s 100 Best Novels                   | modern-librarys-100-best-novels               |             47 | set    |
|         2091 |      38 | The BBC’s 100 Greatest British Novels (2015)       | the-bbcs-100-greatest-british-novels-2015     |             69 | set    |
+--------------+---------+----------------------------------------------------+-----------------------------------------------+----------------+--------+
```

I could change it if you'd prefer:

* Remove the `EbookId` column
* Remove the `SequenceNumber` column
* Make the `UrlName` column unique
* Add a join table, `EbookCollections`, to match what we did with `EbookTags` (with the addition of the `SequenceNumber` column

It would make this read clearer, but a little more work to write collection data.

Reply:
Yes this needs to be two tables:

Collections: CollectionId, Name, UrlName, Type
CollectionEbooks: EbookId, CollectionId, SequenceNumber

Then we can have:

````php
class Ebook{
    public array<Collection> $Collections;
    public function GetSequenceNumber(Collection $collection): ?int{
        // Calculate collection number 
    }
    ...
}

class Collection{
    public array<Ebook> $Ebooks
    ...
}
````

It's not *great* to need a function to get the sequence number but doing it Active Record style would mean creating a subclass which I'm also not a big fan of.

Reply:
Here are my two commits to address this:

https://github.com/standardebooks/web/pull/374/commits/499485366c36feff3fbe81c965039c6ef7a21cd7: Update the DB schema
https://github.com/standardebooks/web/pull/374/commits/a4ce4a73421376de98532ab211e9533a83270ee2: Update the PHP logic

The `Ebook` class already had:

```php

class Ebook{
    public function GetCollectionPosition(Collection $collection): ?int{}
}
```

so I refactored that instead of adding the `Collection::GetSequenceNumber(int $ebookId)` function proposed above. I hope that's ok, because when I tried to do it the other way I ran into some problems when reading the ebook from the filesystem and populating the DB: The `Collection` object doesn't have sequence number property anymore, which is needed to write the ebook to the DB. I tried a few hacks, but I thought they looked worse than what's here.

I was also able to get away with not adding this array of ebooks property:

```php
class Collection{
    public array<Ebook> $Ebooks;
    ...
}
```